walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,117 @@
+import java.util.List;
+import java.util.HashSet;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+      // meeting request has: name, duration, and attendees (List of Strings)
+      // event has: time range, attendees, and name
+
+      long duration = request.getDuration();
+
+       // meeting request exceeds one day
+      if (duration > TimeRange.WHOLE_DAY.duration()) {
+          return Arrays.asList();
+      }
+
+      // all attendees are free
+      if (events.isEmpty()) {
+          return Arrays.asList(TimeRange.WHOLE_DAY);
+      }
+
+      // there are no attendees
+      if (request.getAttendees().isEmpty()) {
+          return Arrays.asList(TimeRange.WHOLE_DAY);
+      }
+      
+
+      Collection<String> mandatory = request.getAttendees();
+      // Collection<String> optional = request.getOptionalAttendees();
+
+      
+      // retrieve all the times when at least one of the mandatory attendees 
+      // is busy
+      List<TimeRange> busy = new ArrayList<TimeRange> ();
+      for (Event e : events) {
+          Set<String> eventAttendees = e.getAttendees();
+          
+          boolean relevantEvent = false;
+          for (String a : eventAttendees) {
+              if (mandatory.contains(a)) {
+                  relevantEvent = true;
+                  break; // only need confirmation that one mandatory attendee is busy
+              }
+          }
+
+          if (relevantEvent) {
+              busy.add(e.getWhen());
+          }
+      }
+
+      // no busy events for mandatory attendees means 
+      // the entire day is free 
+      if (busy.isEmpty()) {
+          return Arrays.asList(TimeRange.WHOLE_DAY);
+      }
+
+      // sort by both end and start time
+      Collections.sort(busy, TimeRange.ORDER_BY_START);
+      ArrayList<TimeRange> busyByEndTime = new ArrayList<TimeRange>(busy);
+      Collections.sort(busyByEndTime, TimeRange.ORDER_BY_END);
+
+      // meeting times that we'll eventually return as suggestion
+      List<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+
+      int startTimeIndex = 0;
+      int endTimeIndex = 0;
+
+      int startTime = TimeRange.START_OF_DAY;
+      int endTime = busy.get(endTimeIndex).start();
+      
+      int numBusyIntervals = busy.size();
+      TimeRange latestMeeting = busyByEndTime.get(numBusyIntervals-1);
+      int lastBusyTime = latestMeeting.end();
+
+      // handle beginning of the day before busy intervals begin
+      TimeRange possibleTime = TimeRange.fromStartEnd(startTime, endTime, false);
+      if(possibleTime.duration() >= duration) {
+            meetingTimes.add(possibleTime);
+      }
+      endTimeIndex++;
+
+      // find all free times between scheduled events
+      while (startTimeIndex < numBusyIntervals && endTimeIndex < numBusyIntervals) {
```

Comment:
Alternatively, you can also just iterate through the busy intervals and compare current_range.end() with (current_range+1).start() 

E.g.

for (int curRangeIdx; curRangeIdx < busy.size()-1; curRangeIdx++) {
  if duration between current_range.end() and (current_range+1).start() >= requested duration
     add to our possible meeting times
}

Reply:
This is to account for the case where one event "contains" several other
ones, like as follows:
…

-------------------------------
|                                    |
.....................................
|.....|   |......|                       |.............|

not a great diagram but I hope it explains the point!